### PR TITLE
Add shader guide

### PR DESCRIPTION
**Pull request by kasenvr**
v1.0 of shader docs, fairly feature complete. Plenty of time in the future to keep optimizing.

### CODE REVIEW

**Comment by odysseus654** [comment]
Do we want to clarify more strongly which shaders do not need enabling? Unless you read that one sentence carefully it sounds like all shaders are default-disabled.

File: docs/source/create/materials/procedural-shaders.rst [hunk 1]
```diff
@@ -0,0 +1,387 @@
+#########################################
+Procedural Shaders for Models and Avatars
+#########################################
+
+You can combine shaders with material entities on shape and zone entities to apply shaders to models and avatars. This feature is currently an **experimental feature** and is under community testing. If you wish to try applying procedural materials to models or avatars *at your own risk*, read on for more information.
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: I think this section needs to be clarified

File: docs/source/create/materials/procedural-shaders.rst [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+Another consideration is that Project Athena does not expose the full range of what GLSL offers to scripters. Instead, users are offered a subset of GLSL, and a custom set of naming conventions for fragment or vertex components.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Basic Method for Using Shaders
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Shaders are stored as a file. Fragment shaders will typically have the file extension ***.fs** or ***.frag** and vertex shaders will usually have the ***.vs** or ***.vert** file extension.
+
+Although you can associate these files with most entities, the original author of Project Athena’s end user shader implementation viewed material entities as the ideal way to apply them to other entities. As a result, this documentation will focus on using shaders with material entities.
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: “associate these files with most entities” doesn’t make sense

File: docs/source/create/materials/procedural-shaders.rst [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+If you are unfamiliar with material entities, you can find more information `here <https://docs.projectathena.dev/create/entities/material-entity.html>`_.
+
+Material entities have data that is stored in the JSON format, and they have a property that is called ``materialData``. The ``materialData`` property will require model and procedural fields. Here is an example::
+
+    {
+        "materials": [{
+            "model": "hifi_shader_simple",
+            "procedural": {
+              "version": 3,
+              "fragmentShaderURL": "https://gist.githubusercontent.com/MarcusLlewellyn/2726885d4cc8be086a6755f05fbd99ff/raw/rotatecube.fs"
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: need to replace this with a real v3 shader, ideally in our repo

File: docs/source/create/materials/procedural-shaders.rst [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+^^^^^^^^^^^^^^^
+Shader Template
+^^^^^^^^^^^^^^^
+
+When you learn about shaders for other applications, the shader may have a function like ``main()`` that is run first. By contrast, Project Athena has a specific function name that must be called. Which function is used depends on which version of the shader you use.
+
+As shaders were developed, features for them evolved a bit over time. As a result, there are several shader versions, and each version has a different call signature. **Versions 1 and 2** are the oldest, and will still work. **Versions 3 and 4** are the newest and expose more features. Version 4 provides for per-fragment positions, however it is also the most expensive. Therefore it is recommended to use Version 3 if that extra feature from Version 4 is not needed.
+
+The most basic template for a shader will look something like this example::
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: this example should be a v3 shader

File: docs/source/create/materials/procedural-shaders.rst [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+    const vec4 iChannelTime = vec4(0.0); // No support for video input
+
+The following per-fragment uniforms are also provided in all shader versions::
+
+    vec4 _positionMS; (equal to _position)
+    vec4 _positionES; (equal to _eyePosition)
+    vec3 _normalMS; (equal to _modelNormal)
+    vec3 _normalWS; (equal to _normal)
+    vec4 _color;
+    vec4 _texCoord01 (also split into vec2_texCoord0 and vec2 _texCoord1)
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: space after vec2 before _texCoord0

File: docs/source/create/materials/procedural-shaders.rst [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+    // https://en.wikipedia.org/wiki/Fractional_Brownian_motion
+    float hifi_fbm(in vec2 p);
+    float DEFAULT_ROUGHNESS = 0.9;
+    float DEFAULT_SHININESS = 10.0;
+    float DEFAULT_METALLIC = 0.0;
+    vec3 DEFAULT_SPECULAR = vec3(0.1);
+    vec3 DEFAULT_EMISSIVE = vec3(0.0);
+    float DEFAULT_OCCLUSION = 1.0;
+    float DEFAULT_SCATTERING = 0.0;
+    vec3 DEFAULT_FRESNEL = DEFAULT_EMISSIVE;
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: these are already listed above

File: docs/source/create/materials/procedural-shaders.rst [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+    float isBloomEnabled()
+    float isSkinningEnabled()
+    float isBlendshapeEnabled()
+    
+Shader Version 1
+----------------
+::
+    
+    // Must implement. Always emissive, returns a single color.
+    uniform vec3 getProceduralColor()
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: remove uniform

File: docs/source/create/materials/procedural-shaders.rst [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+        vec3 emissive;
+        float alpha;
+        float roughness;
+        float metallic;
+        float occlusion;
+        float scattering;
+    };
+
+This is the same as Shader Version 3 but with per-fragment position. By modifying position, you can modify the per-fragment depth. This allows you to create things like ray-marched geometry that depth-tests properly and is dynamically lit by light entities. The trade-off is that this version is much more expensive than Version 3.
+
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: there’s no section on zones here, see my comment on the original issue

File: docs/source/create/materials/procedural-shaders.rst (hunk 4 above)
Comment (HifiExperiments) [reply docs/source/create/materials/procedural-shaders.rst]: also two : here

Comment (HifiExperiments) [reply docs/source/create/materials/procedural-shaders.rst]: and elsewhere

Comment (kasenvr) [reply docs/source/create/materials/procedural-shaders.rst]: I will need a very simple example that can be used as a starting point for a v3 shader to replace that. :)

**Comment by kasenvr** [comment]
https://docs.projectathena.dev/staging/html/create/materials/procedural-shaders.html

The latest and greatest preview for the docs.

**Comment by odysseus654** [comment]
Here's all the shader versions I can think of.  All but the last were tested on a new cube (didn't test the skybox one, but it was based on a previously-written shader):

```GLSL
 // version 1
 vec3 getProceduralColor() {
 	return _positionMS.xyz;
 }
 
 // version 2
 float getProceduralColors(inout vec3 diffuse, inout vec3 specular, inout float shininess) {
 	// diffuse is from the texture, others are hardcoded to DEFAULT_SPECULAR and DEFAULT_SHININESS
 	diffuse = _positionMS.xyz;
 	return 1.0; // emissive, between 0.0 - 1.0
 }
 
 // version 3
 float getProceduralFragment(inout ProceduralFragment data) {
 	data.diffuse = vec3(0);
 	data.occlusion = 0;
 	data.roughness = 1;
 	data.emissive = _positionMS.xyz;
 	return 0; // "emissiveAmount", either <=0 or >0, suggest return 0 and use data.emissive
 }
 
 // version 4
 float getProceduralFragmentWithPosition(inout ProceduralFragmentWithPosition data) {
 	data.diffuse = vec3(0);
 	data.occlusion = 0;
 	data.roughness = 1;
 	data.emissive = _positionMS.xyz;
 	return 0; // "emissiveAmount", either <=0 or >0, suggest return 0 and use data.emissive
 }

 // skybox
 vec3 getSkyboxColor() {
 	vec3 normal = normalize(_normal);
 	return texture(cubeMap, normal).rgb; // this should return the same value that the skybox texture has
 }
```

**Comment by odysseus654** [comment]
Descriptions (mostly guesses) based on playing with shaders:

vec4 _positionMS; // position in "model space" (relative to the center of the object)
vec4 _positionES; // position in "eye space" (relative to the center of your eye)
vec3 _normalMS; // direction the current face is pointing in "model space" (without any rotations)
vec3 _normalWS; // direction the current face is pointing in "world space" (after rotations applied)
vec4 _color; // color of the object
vec2 _texCoord0 // UV texture coordinates on this model
vec2 _texCoord1 // second UV texture coordinates?!?

**Comment by odysseus654** [comment]
bool cam_isStereo() // is user wearing a VR headset (or a 3D monitor?)
float cam_getStereoSide() // 1 for right eye in a stereo context, otherwise 0

**Comment by kasenvr** [comment]
The latest version can be seen here:

https://docs.projectathena.dev/staging/html/create/materials/procedural-shaders.html

File: docs/source/create/materials/procedural-shaders.rst [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+Shaders are applied to entities and avatars by way of attaching them to a material. That material is then attached to your entity or avatar, therefore applying the shader as intended.
+
+If you are unfamiliar with material entities, you can find more information `here <https://docs.projectathena.dev/create/entities/material-entity.html>`_.
+
+Material entities have data that is stored in the JSON format, and they have a property that is called ``materialData``. The ``materialData`` property will require model and procedural fields. Here is an example::
+
+    {
+        "materials": [{
+            "model": "hifi_shader_simple"
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: need a comma at end of line

File: docs/source/create/materials/procedural-shaders.rst [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+
+The ``materialData`` JSON can be applied either via the Project Athena Interface’s edit tools or with a script. Here's another example::
+
+    Entities.addEntity({
+    	type: "Material",
+    	materialURL: "materialData",
+    	priority: 1,
+    	materialData: JSON.stringify({
+    		materials: {
+    			"model": "hifi_shader_simple"
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: need a comma at end of line

File: docs/source/create/materials/procedural-shaders.rst [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+    // version 3
+    float getProceduralFragment(inout ProceduralFragment data) {
+        data.diffuse = vec3(0);
+        data.occlusion = 0;
+        data.roughness = 1;
+        data.emissive = _positionMS.xyz;
+        return 0; // "emissiveAmount", either <=0 or >0, suggest return 0 and use data.emissive
+    }
+
+The function ``getProceduralFragmentWithPosition()`` is the default main entry point for the fragment shader. Because shaders are always read by their compiler from top to bottom, this function must always be the last one in your shader code.
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: most of this paragraph and the next are already covered below (also they refer to v4 things but the example is v3)

File: docs/source/create/materials/procedural-shaders.rst [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+    const float DEFAULT_METALLIC = 0.0;
+    const vec3 DEFAULT_SPECULAR = vec3(0.1);
+    const vec3 DEFAULT_EMISSIVE = vec3(0.0);
+    const float DEFAULT_OCCLUSION = 1.0;
+    const float DEFAULT_SCATTERING = 0.0;
+    const vec3 DEFAULT_FRESNEL = DEFAULT_EMISSIVE;
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+Shader Examples by Version
+^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: I don’t think this section is necessary.  each version is already explained below.  perhaps just an examples section at the bottom of the whole page?

File: docs/source/create/materials/procedural-shaders.rst [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+    const vec3 iResolution = vec3(1.0); // Resolution doesn’t make sense in the VR context
+    const vec4 iMouse = vec4(0.0); // Mouse functions not enabled currently
+    const float iSampleRate = 1.0; // No support for audio input
+    const vec4 iChannelTime = vec4(0.0); // No support for video input
+
+The following per-fragment uniforms are also provided in all shader versions::
+
+    vec4 _positionMS; // position in "model space" (relative to the center of the object); (equal to _position)
+    vec4 _positionES; // position in "eye space" (relative to the center of your eye); (equal to _eyePosition)
+    vec3 _normalMS; direction the current face is pointing in "model space" (without any rotations); (equal to _modelNormal)
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: missing // for comment

File: docs/source/create/materials/procedural-shaders.rst [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+        vec3 normal;
+        vec3 nonSkinnedNormal; // input only
+        vec3 tangent; // input only
+        vec3 nonSkinnedTangent; // input only
+        vec4 color;
+        vec2 texCoord0;
+    };
+
+--------------------------------------
+For Both Procedural and Vertex Shaders
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: for both fragment and vertex shaders

File: docs/source/create/materials/procedural-shaders.rst (hunk 11 above)
Comment (kasenvr) [reply docs/source/create/materials/procedural-shaders.rst]: https://docs.projectathena.dev/staging/html/create/materials/procedural-shaders.html#shader-template Here is the revision.

File: docs/source/create/materials/procedural-shaders.rst [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+
+To set the materialData using the edit tools, you will want to set ``materialURL`` equal to **materialData**. Then you will want to put in the JSON.stringify'd version of the materialData into the field::
+
+    materials: {
+        "model": "hifi_shader_simple",
+        "procedural": {
+            "version": 3,
+            "shaderUrl": "https://docs.projectathena.dev/_static/resources/Proceduralv3.fs"
+        }
+    }
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: this was already explained above

File: docs/source/create/materials/procedural-shaders.rst [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+The default values for some of these are::
+
+    const float DEFAULT_ROUGHNESS = 0.9;
+    const float DEFAULT_SHININESS = 10.0;
+    const float DEFAULT_METALLIC = 0.0;
+    const vec3 DEFAULT_SPECULAR = vec3(0.1);
+    const vec3 DEFAULT_EMISSIVE = vec3(0.0);
+    const float DEFAULT_OCCLUSION = 1.0;
+    const float DEFAULT_SCATTERING = 0.0;
+    const vec3 DEFAULT_FRESNEL = DEFAULT_EMISSIVE;
```
Comment (HifiExperiments) [inline docs/source/create/materials/procedural-shaders.rst]: these should probably be in the v3 section above

File: docs/source/create/materials/procedural-shaders.rst [hunk 17]
```diff
@@ -0,0 +1,434 @@
...
+
+To set the materialData using the edit tools, you will want to set ``materialURL`` equal to **materialData**. Then put in the JSON.stringify'd version of the materialData as shown above into the field.
+
+^^^^^^^^^^^^^^^
+Shader Template
+^^^^^^^^^^^^^^^
+
+When you learn about shaders for other applications, the shader may have a function like ``main()`` that is run first. By contrast, Project Athena has a specific function name that must be called. Which function is used depends on which version of the shader you use.
+
+As shaders were developed, features for them evolved a bit over time. As a result, there are several shader versions, and each version has a different call signature. **Versions 1 and 2** are the oldest, and will still work. **Versions 3 and 4** are the newest and expose more features. Version 4 provides for per-fragment positions, however it is also the most expensive. Therefore it is recommended to use Version 3 if that extra feature from Version 4 is not needed.
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
As shaders were developed, features for them evolved a bit over time. As a result, there are several shader versions, and each version has a different call signature. **Versions 1 and 2** are the oldest, and will still work. **Versions 3 and 4** are the newest and expose more features. Version 4 provides for per-fragment positions, however it is also the most computationally expensive. Therefore it is recommended to use version 3 if that extra feature from version 4 is not needed.
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 18]
```diff
@@ -0,0 +1,434 @@
...
+
+A shader consists of two primary pieces, **the main function** that is responsible for coloring the pixel and then any desired **helper functions** that assist in that processing logic, which will go above the main function.
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
A shader consists of two primary pieces: **the main function** that is responsible for coloring the pixel, and any desired **helper functions** that assist in that processing logic (which must go above the main function).
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 19]
```diff
@@ -0,0 +1,434 @@
...
+
+A basic template for a shader without helper functions will look something like this example::
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
A basic template for a shader without helper functions looks something like this example::
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 20]
```diff
@@ -0,0 +1,434 @@
...
+        vec3 specular;
+        vec3 emissive;
+        float alpha;
+        float roughness;
+        float metallic;
+        float occlusion;
+        float scattering;
+    };
+
+This is the same as Shader Version 3 but with per-fragment position. By modifying position, you can modify the per-fragment depth. This allows you to create things like ray-marched geometry that depth-tests properly and is dynamically lit by light entities. The trade-off is that this version is much more expensive than Version 3.
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
This is the same as shader version 3 but with per-fragment position. By modifying position, you can modify the per-fragment depth. This allows you to create things like ray-marched geometry that depth-tests properly and is dynamically lit by light entities. The trade-off is that this version is much more computationally expensive than version 3.
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 21]
```diff
@@ -0,0 +1,434 @@
...
+
+^^^^^^^^^^^^^
+Zone Entities
+^^^^^^^^^^^^^
+
+Zones operate slightly differently. They support the same global defines, but not the provided methods or constants. They also provide the following inputs:
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
Zones entities operate slightly differently. They support the same global defines but not the provided methods or constants. They also provide the following inputs:
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 22]
```diff
@@ -0,0 +1,434 @@
...
+::
+
+    vec3 _normal;
+    Skybox skybox; (a struct containing vec4 color)
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
    Skybox skybox; // a struct containing vec4 color
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 23]
```diff
@@ -0,0 +1,434 @@
...
+    samplerCube cubeMap; (the skybox texture)
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
    samplerCube cubeMap; // the skybox texture
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 24]
```diff
@@ -0,0 +1,434 @@
...
+        vec4 color;
+        vec2 texCoord0;
+    };
+
+--------------------------------------
+For Both Fragment and Vertex Shaders
+--------------------------------------
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Custom uniforms and textures
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
Custom Uniforms and Textures
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 25]
```diff
@@ -0,0 +1,434 @@
...
+When texture URLs are provided, iChannel0 - iChannel3 will be populated, as well as iChannelResolution[0] - iChannelResolution[3].
+
+When you provide uniforms, you must also include them at the top of your shader file, with optional defaults::
+
+    uniform vec3 _diffuse = vec3(0.0);
+    uniform float _alpha = 1.0;
+    uniform vec3 _emissive = vec3(0.0);
+    uniform float _emissiveAmount = 0.0;
+
+Supported uniform types are: ``float``, ``vec2``, ``vec3``, and ``vec4`` (multiple values are provided as arrays.)
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
Supported uniform types are: ``float``, ``vec2``, ``vec3``, and ``vec4``. (Multiple values are provided as arrays.)
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 26]
```diff
@@ -0,0 +1,434 @@
...
+Alpha Effects (Transparency)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Shaders that make use of the ``proceduralData.alpha`` value won’t display alpha on their own. In order for a shader’s alpha to be active, the entity it is applied to must first have either its alpha property less than ``1.0``, or a material property setting opacity to less than ``1.0``.
+
+^^^^^^^^^^^^^^^^^
+Debugging Shaders
+^^^^^^^^^^^^^^^^^
+
+The only way to debug shaders at the moment is to look at the interface’s log file. Shader compilation errors will appear in this log, and can help with locating issues.
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
The only way to debug shaders at the moment is to look at Interface’s log file. Shader compilation errors will appear in this log, and can help with locating issues.
```

File: docs/source/create/materials/procedural-shaders.rst [hunk 27]
```diff
@@ -286,18 +286,18 @@ Shader Version 4
...
 
-This is the same as Shader Version 3 but with per-fragment position. By modifying position, you can modify the per-fragment depth. This allows you to create things like ray-marched geometry that depth-tests properly and is dynamically lit by light entities. The trade-off is that this version is much more expensive than Version 3.
+This is the same as shader version 3 but with per-fragment position. By modifying position, you can modify the per-fragment depth. This allows you to create things like ray-marched geometry that depth-tests properly and is dynamically lit by light entities. The trade-off is that this version is much more computationally expensive than version 3.
 
 ^^^^^^^^^^^^^
 Zone Entities
 ^^^^^^^^^^^^^
 
-Zones operate slightly differently. They support the same global defines, but not the provided methods or constants. They also provide the following inputs:
+Zones entities operate slightly differently. They support the same global defines but not the provided methods or constants. They also provide the following inputs:
```
Comment (ctrlaltdavid) [inline docs/source/create/materials/procedural-shaders.rst]:
```suggestion
Zone entities operate slightly differently. They support the same global defines but not the provided methods or constants. They also provide the following inputs:
```

**Pull request merged**